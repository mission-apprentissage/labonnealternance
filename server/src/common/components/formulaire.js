import moment from "moment"
import { ANNULEE, POURVUE } from "../constants.js"
import { Formulaire } from "../model/index.js"

export default () => ({
  getFormulaires: async (query, options, { page, limit }) => {
    const response = await Formulaire.paginate(query, { ...options, page, limit, lean: true })
    return {
      pagination: {
        page: response.page,
        result_per_page: limit,
        number_of_page: response.pages,
        total: response.total,
      },
      data: response.docs,
    }
  },
  getFormulaire: (query) => Formulaire.findOne(query).lean(),
  createFormulaire: (payload) => Formulaire.create(payload),
  deleteFormulaire: (formId) => Formulaire.findByIdAndDelete(formId),
  deleteFormulaireFromGestionnaire: (siret) => Formulaire.deleteMany({ gestionnaire: siret }),
  updateFormulaire: (id_form, payload) => Formulaire.findOneAndUpdate({ id_form }, payload, { new: true }),
  archiveFormulaire: async (id_form) => {
    let form = await Formulaire.findOne({ id_form })

    form.offres.map((offre) => {
      offre.statut = "Annulée"
    })

    form.statut = "Archivé"
    await form.save()

    return true
  },
  getOffre: (id) => Formulaire.findOne({ "offres._id": id }).lean(),
  createOffre: (id_form, payload) => Formulaire.findOneAndUpdate({ id_form }, { $push: { offres: payload } }, { new: true }),
  updateOffre: (id_offre, payload) =>
    Formulaire.findOneAndUpdate(
      { "offres._id": id_offre },
      {
        $set: {
          "offres.$": payload,
        },
      },
      { new: true }
    ),
  provideOffre: async (id_offre) => {
    await Formulaire.findOneAndUpdate(
      { "offres._id": id_offre },
      {
        $set: {
          "offres.$.statut": POURVUE,
        },
      }
    )
    return true
  },
  cancelOffre: async (id_offre) => {
    await Formulaire.findOneAndUpdate(
      { "offres._id": id_offre },
      {
        $set: {
          "offres.$.statut": ANNULEE,
        },
      }
    )
    return true
  },
  extendOffre: async (id_offre) => {
    await Formulaire.findOneAndUpdate(
      { "offres._id": id_offre },
      {
        $set: {
          "offres.$.date_expiration": moment().add(1, "months").format("YYYY-MM-DD"),
          "offres.$.date_derniere_prolongation": Date.now(),
        },
        $inc: { "offres.$.nombre_prolongation": 1 },
      }
    )
    return true
  },
})
