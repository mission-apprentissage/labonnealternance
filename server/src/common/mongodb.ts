import mongoose from "mongoose"
import config from "../config.js"

export let mongooseInstance = mongoose
export const connectToMongo = (mongoUri = config.mongodb.uri, mongooseInst = null) => {
  return new Promise((resolve, reject) => {
    console.log(`MongoDB: Connection to ${mongoUri}`)

    const mI = mongooseInst || mongooseInstance
    // Set up default mongoose connection
    mI.connect(mongoUri, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
      useFindAndModify: false,
      keepAlive: true,
    })

    // Get Mongoose to use the global promise library
    mI.Promise = global.Promise // Get the default connection
    const db = mI.connection

    // Bind connection to error event (to get notification of connection errors)
    db.on("error", (e) => {
      console.error("MongoDB: connection error:")
      reject(e)
    })

    db.once("open", () => {
      console.log("MongoDB: Connected")
      resolve({ db })
    })
  })
}

export const closeMongoConnection = (mongooseInst = mongoose) => mongooseInst.disconnect()
