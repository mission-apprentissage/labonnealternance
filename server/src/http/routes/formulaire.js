import express from "express"
import { mailTemplate } from "../../assets/index.js"
import { getCatalogueEtablissements, getCatalogueFormations } from "../../common/catalogue.js"
import dayjs from "../../common/dayjs.js"
import { getElasticInstance } from "../../common/esClient/index.js"
import { Formulaire, UserRecruteur } from "../../common/model/index.js"
import config from "../../config.js"
import { tryCatch } from "../middlewares/tryCatchMiddleware.js"

const esClient = getElasticInstance()

export default ({ formulaire, mailer, etablissementsRecruteur, application, usersRecruteur }) => {
  const router = express.Router()

  /**
   * Query search endpoint
   */
  router.get(
    "/",
    tryCatch(async (req, res) => {
      let query = JSON.parse(req.query.query)
      const results = await Formulaire.find(query).lean()

      return res.json(results)
    })
  )

  /**
   * Get form from id
   *
   */
  router.get(
    "/:id_form",
    tryCatch(async (req, res) => {
      let result = await formulaire.getFormulaire({ id_form: req.params.id_form })

      if (!result) {
        return res.sendStatus(401)
      }

      await Promise.all(
        result.offres.map(async (offre) => {
          let candidatures = await application.getApplication(offre._id)

          if (candidatures) {
            offre.candidatures = candidatures.length > 0 ? candidatures.length : undefined
          }

          return offre
        })
      )

      return res.json(result)
    })
  )

  /**
   * Post form
   */
  router.post(
    "/",
    tryCatch(async (req, res) => {
      const response = await formulaire.createFormulaire(req.body)

      /**
       * HOTFIX 18/02 : dans le modèle précédent du widget, les utilisateurs n'était pas créé dans la collection UserRecruteur.
       */
      if (req.body.origine === "akto") {
        let exist = await usersRecruteur.getUser({ email: req.body.email })

        if (!exist) {
          await usersRecruteur.createUser({
            ...req.body,
            type: "ENTREPRISE",
            id_form: response.id_form,
            email_valide: true,
          })
        }
      }

      return res.json(response)
    })
  )

  /**
   * Put form
   */
  router.put(
    "/:id_form",
    tryCatch(async (req, res) => {
      const result = await formulaire.updateFormulaire(req.params.id_form, req.body)
      return res.json(result)
    })
  )

  router.delete(
    "/:id_form",
    tryCatch(async (req, res) => {
      await formulaire.archiveFormulaire(req.params.id_form)
      return res.sendStatus(200)
    })
  )

  /**
   * TEMP : get offer from id for front
   * TO BE REPLACE
   */
  router.get(
    "/offre/f/:id_offre",
    tryCatch(async (req, res) => {
      const result = await formulaire.getOffre(req.params.id_offre)
      const offre = result.offres.filter((x) => x._id == req.params.id_offre)

      res.json(offre)
    })
  )

  /**
   * LBA ENDPOINT : get offer from id
   *
   * 25/04/2022 : to export to specific route file with auth for LBA endpoint
   */
  router.get(
    "/offre/:id_offre",
    tryCatch(async (req, res) => {
      const result = await formulaire.getOffre(req.params.id_offre)

      if (!result) {
        return res.status(400).json({ error: true, message: "Not found" })
      }

      result.offres = result.offres.filter((x) => x._id == req.params.id_offre)

      if (result.mandataire === true) {
        const [entreprise_localite] = result.adresse.match(/([0-9]{5})[ ,] ?([A-zÀ-ÿ]*)/) ?? [""]
        const cfa = await etablissementsRecruteur.getEtablissement({ siret: result.gestionnaire })

        result.telephone = cfa.telephone
        result.email = cfa.email
        result.nom = cfa.nom
        result.prenom = cfa.prenom
        result.raison_sociale = cfa.raison_sociale
        result.adresse = cfa.adresse
        result.entreprise_localite = entreprise_localite
      }

      result.events = undefined
      result.mailing = undefined

      return res.json(result)
    })
  )

  /**
   * Create new offer
   */

  router.post(
    "/:id_form/offre",
    tryCatch(async (req, res) => {
      const result = await formulaire.createOffre(req.params.id_form, req.body)

      let { email, raison_sociale, prenom, nom, mandataire, gestionnaire, offres } = result
      let offre = req.body
      let contactCFA

      offre._id = result.offres.filter((x) => x.libelle === offre.libelle)[0]._id

      offre.supprimer = `${config.publicUrlEspacePro}/offre/${offre._id}/cancel`
      offre.pourvue = `${config.publicUrlEspacePro}/offre/${offre._id}/provided`

      // if first offer creation for an Entreprise, send specific mail
      if (offres.length === 1 && mandataire === false) {
        // Get user account
        const user = await usersRecruteur.getUser({ email })
        // Get user account validation link
        const url = etablissementsRecruteur.getValidationUrl(user._id)

        await mailer.sendEmail({
          to: email,
          subject: "La bonne alternance - Merci de valider votre adresse mail pour diffuser votre offre",
          template: mailTemplate["mail-nouvelle-offre-depot-simplifie"],
          data: {
            nom: user.nom,
            prenom: user.prenom,
            email: user.email,
            confirmation_url: url,
            offres: [offre],
          },
        })

        return res.json(result)
      }

      // get CFA informations if formulaire is handled by a CFA
      if (result.mandataire) {
        contactCFA = await UserRecruteur.findOne({ siret: gestionnaire })
      }

      // Send mail with action links to manage offers
      await mailer.sendEmail({
        to: mandataire ? contactCFA.email : email,
        subject: mandataire
          ? `La bonne alternance - Votre offre d'alternance pour ${raison_sociale} a bien été publiée`
          : `La bonne alternance - Votre offre d'alternance a bien été publiée`,
        template: mailTemplate["mail-nouvelle-offre"],
        data: {
          nom: mandataire ? contactCFA.nom : nom,
          prenom: mandataire ? contactCFA.prenom : prenom,
          raison_sociale,
          mandataire: result.mandataire,
          offres: [offre],
          lba_url:
            config.env !== "recette"
              ? `https://labonnealternance.apprentissage.beta.gouv.fr/recherche-apprentissage?&display=list&page=fiche&type=matcha&itemId=${offre._id}`
              : `https://labonnealternance-recette.apprentissage.beta.gouv.fr/recherche-apprentissage?&display=list&page=fiche&type=matcha&itemId=${offre._id}`,
        },
      })

      return res.json(result)
    })
  )

  /**
   * Create offer delegations
   */
  router.post(
    "/offre/:idOffre/delegation",
    tryCatch(async (req, res) => {
      const { etablissementCatalogueIds } = req.body
      const { idOffre } = req.params

      const offreDocument = await formulaire.getOffre(idOffre)
      const offre = offreDocument.offres.find((offre) => offre._id.toString() === idOffre)

      const { etablissements } = await getCatalogueEtablissements({ _id: { $in: etablissementCatalogueIds } })

      const promises = etablissements.map(async (etablissement) => {
        const { formations } = await getCatalogueFormations(
          {
            $or: [
              {
                etablissement_gestionnaire_id: etablissement._id,
              },
              {
                etablissement_formateur_id: etablissement._id,
              },
            ],
            etablissement_gestionnaire_courriel: { $nin: [null, ""] },
            catalogue_published: true,
          },
          { etablissement_gestionnaire_courriel: 1 }
        )

        await mailer.sendEmail({
          to: formations[0].etablissement_gestionnaire_courriel,
          subject: `Une entreprise recrute dans votre domaine`,
          template: mailTemplate["mail-cfa-delegation"],
          data: {
            enterpriseName: offreDocument.raison_sociale,
            jobName: offre.rome_appellation_label,
            contractType: offre.type[0],
            trainingLevel: offre.niveau,
            startDate: dayjs(offre.date_debut_apprentissage).format("DD/MM/YYYY"),
            duration: offre.duree_contrat,
            rhythm: offre.rythme_alternance,
            offerButton: `${config.publicUrlEspacePro}/proposition/formulaire/${offreDocument.id_form}/offre/${offre._id}`,
            createAccountButton: `${config.publicUrlEspacePro}/creation/cfa`,
          },
        })
      })

      await Promise.all(promises)

      return res.json(offre)
    })
  )

  /**
   * Put existing offer from id
   */
  router.put(
    "/offre/:id_offre",
    tryCatch(async (req, res) => {
      const result = await formulaire.updateOffre(req.params.id_offre, req.body)
      return res.json(result)
    })
  )

  /**
   * Permet de passer une offre en statut ANNULER (mail transactionnel)
   */
  router.put(
    "/offre/:offreId/cancel",
    tryCatch(async (req, res) => {
      const exist = formulaire.getOffre(req.params.offreId)

      if (!exist) {
        return res.status(400).json({ status: "INVALID_RESSOURCE", message: "Offre does not exist" })
      }

      await formulaire.cancelOffre(req.params.offreId)

      return res.sendStatus(200)
    })
  )

  /**
   * Permet de passer une offre en statut POURVUE (mail transactionnel)
   */
  router.put(
    "/offre/:offreId/provided",
    tryCatch(async (req, res) => {
      const exist = formulaire.getOffre(req.params.offreId)

      if (!exist) {
        return res.status(400).json({ status: "INVALID_RESSOURCE", message: "Offre does not exist" })
      }

      await formulaire.provideOffre(req.params.offreId)

      return res.sendStatus(200)
    })
  )

  /**
   * LBA ENDPOINT
   */
  router.post(
    "/search",
    tryCatch(async (req, res) => {
      const { distance, lat, lon, romes, niveau } = req.body
      const filter = [
        {
          geo_distance: {
            distance: `${distance}km`,
            geo_coordonnees: {
              lat,
              lon,
            },
          },
        },
      ]

      if (!distance || !lat || !lon || !romes) {
        return res.status(400).json({ error: "Argument is missing (distance, lat, lon, romes)" })
      }

      if (niveau && niveau !== "Indifférent") {
        filter.push({
          nested: {
            path: "offres",
            query: {
              bool: {
                must: [
                  {
                    match: {
                      "offres.niveau": {
                        query: niveau,
                        fuzziness: 0,
                        operator: "AND",
                      },
                    },
                  },
                ],
              },
            },
          },
        })
      }

      let body = {
        query: {
          bool: {
            must: [
              {
                nested: {
                  path: "offres",
                  query: {
                    bool: {
                      must: [
                        {
                          match: {
                            "offres.romes": romes.join(" "),
                          },
                        },
                        {
                          match: {
                            "offres.statut": "Active",
                          },
                        },
                      ],
                    },
                  },
                },
              },
            ],
            filter: filter,
          },
        },
        sort: [
          {
            _geo_distance: {
              geo_coordonnees: {
                lat,
                lon,
              },
              order: "asc",
              unit: "km",
              mode: "min",
              distance_type: "arc",
              ignore_unmapped: true,
            },
          },
        ],
      }

      const result = await esClient.search({ index: "formulaires", body })

      const filtered = await Promise.all(
        result.body.hits.hits.map(async (x) => {
          let offres = []
          let cfa = {}

          if (x._source.offres.length === 0) {
            return
          }

          x._source.mailing = undefined
          x._source.events = undefined

          if (x._source.mandataire === true) {
            const [entreprise_localite] = x._source.adresse.match(/([0-9]{5})[ ,] ?([a-zA-Z-]*)/) ?? [""]
            cfa = await etablissementsRecruteur.getEtablissement({ siret: x._source.gestionnaire })

            x._source.telephone = cfa.telephone
            x._source.email = cfa.email
            x._source.nom = cfa.nom
            x._source.prenom = cfa.prenom
            x._source.raison_sociale = cfa.raison_sociale
            x._source.adresse = cfa.adresse
            x._source.entreprise_localite = entreprise_localite
          }

          x._source.offres.forEach((o) => {
            if (romes.some((item) => o.romes.includes(item)) && o.statut === "Active") {
              o.libelle = o.rome_appellation_label ?? o.libelle
              offres.push(o)
            }
          })

          x._source.offres = offres
          return x
        })
      )

      return res.json(filtered)
    })
  )

  return router
}
