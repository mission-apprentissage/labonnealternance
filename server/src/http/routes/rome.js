import axios from "axios"
import express from "express"
import querystring from "querystring"
import dayjs from "../../common/dayjs.js"
import config from "../../config.js"
import { getRomesAndLabelsFromTitleQuery } from "../../service/domainesMetiers.js"
import { tryCatch } from "../middlewares/tryCatchMiddleware.js"

/**
 * API romes
 */

const isTokenValid = (token) => token.expire?.isAfter(dayjs())

const getToken = async (token = {}) => {
  let isValid = isTokenValid(token)

  if (isValid) {
    return token
  }

  try {
    const response = await axios.post(
      "https://entreprise.pole-emploi.fr/connexion/oauth2/access_token?realm=partenaire",
      querystring.stringify({
        grant_type: "client_credentials",
        client_id: config.poleEmploi.clientId,
        client_secret: config.poleEmploi.clientSecret,
        scope: `api_romev1 application_${config.poleEmploi.clientId} nomenclatureRome`,
      }),
      {
        headers: {
          Authorization: `Bearer ${config.poleEmploi.clientSecret}`,
          "Content-Type": "application/x-www-form-urlencoded",
        },
      }
    )

    return {
      ...response.data,
      expire: dayjs().add(response.data.expires_in - 10, "s"),
    }
  } catch (error) {
    console.log(error)
    return error.response.data
  }
}

export default function () {
  const router = express.Router()
  let token = {}

  router.get(
    "/",
    tryCatch(async (req, res) => {
      const result = await getRomesAndLabelsFromTitleQuery(req.query)
      return res.json(result)
    })
  )

  router.get(
    "/detail/:rome",
    tryCatch(async (req, res) => {
      token = await getToken(token)

      let response = await axios.get(`https://api.pole-emploi.io/partenaire/rome/v1/metier/${req.params.rome}`, {
        headers: {
          Authorization: `Bearer ${token.access_token}`,
        },
      })

      return res.json(response.data)
    })
  )

  return router
}
