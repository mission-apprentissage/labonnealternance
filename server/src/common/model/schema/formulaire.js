import { nanoid } from "nanoid"
import { mongooseInstance } from "../../mongodb.js"
import { offreSchema } from "./offre.js"

export const formulaireSchema = mongooseInstance.Schema(
  {
    id_form: {
      type: String,
      default: () => nanoid(),
      description: "Identifiant de formulaire unique",
      index: true,
    },
    raison_sociale: {
      type: String,
      default: null,
      description: "Raison social de l'entreprise",
    },
    enseigne: {
      type: String,
      default: null,
      description: "Enseigne de l'entreprise",
    },
    siret: {
      type: String,
      default: null,
      description: "Numéro SIRET de l'entreprise",
    },
    adresse: {
      type: String,
      default: null,
      description: "Adresse de l'entreprise",
    },
    geo_coordonnees: {
      type: String,
      default: null,
      description: "Latitude/Longitude (inversion lié à LBA) de l'adresse de l'entreprise",
    },
    mandataire: {
      type: Boolean,
      default: false,
      description: "le formulaire est-il géré par un mandataire ?",
    },
    gestionnaire: {
      type: String,
      description: "Siret de l'organisme de formation gestionnaire des offres de l'entreprise",
    },
    nom: {
      type: String,
      default: null,
      description: "Nom du contact",
    },
    prenom: {
      type: String,
      default: null,
      description: "Prénom du contact",
    },
    telephone: {
      type: String,
      default: null,
      description: "Téléphone du contact",
    },
    email: {
      type: String,
      default: null,
      description: "Email du contact",
    },
    offres: [{ type: offreSchema, default: {}, description: "Liste des offres d'apprentissage" }],
    origine: {
      type: String,
      default: null,
      description: "Origine/organisme lié au formulaire",
    },
    opco: {
      type: String,
      description: "Information sur l'opco de l'entreprise",
    },
    idcc: {
      type: String,
      description: "Identifiant convention collective de l'entreprise",
    },
    statut: {
      type: String,
      enum: ["Actif", "Archivé", "En attente de validation"],
      default: "Actif",
      description: "Statut du formulaire",
    },
    code_naf: {
      type: String,
      description: "code NAF de l'entreprise",
    },
    libelle_naf: {
      type: String,
      description: "Libelle du code NAF de l'entreprise",
    },
    tranche_effectif: {
      type: String,
      description: "Tranche d'effectif salariale de l'entreprise",
    },
    date_creation_etablissement: {
      type: Date,
      description: "Date de creation de l'entreprise",
    },
  },
  {
    timestamps: true,
  }
)
