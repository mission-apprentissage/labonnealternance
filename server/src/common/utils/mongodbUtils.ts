import { captureException } from "@sentry/node"
import { isEqual } from "lodash-es"
import { Collection, CollectionInfo, MongoClient, MongoServerError } from "mongodb" // to uncomment when migrated to V7
import { IModelDescriptor } from "shared/models/common"
import { CollectionName, IDocument, modelDescriptors } from "shared/models/models"
import { zodToMongoSchema } from "zod-mongodb-schema"

import config from "../../config"
import { logger } from "../logger"

import { sleep } from "./asyncUtils"

let mongodbClient: MongoClient | null = null
let mongodbClientState: string | null = null

export const ensureInitialization = () => {
  if (!mongodbClient) {
    throw new Error("Database connection does not exist. Please call connectToMongodb before.")
  }
  return mongodbClient
}

/**
 * @param  {string} uri
 * @returns client
 */
export const connectToMongodb = async (uri: string) => {
  const client = new MongoClient(uri, {
    heartbeatFrequencyMS: 10_000,
    retryWrites: true,
    retryReads: true,
    minPoolSize: config.env === "local" ? 0 : 5,
    maxPoolSize: 1_000,
  })

  client.on("connectionPoolReady", () => {
    logger.info("MongoDB reconnected")
    mongodbClient = client
  })

  client.on("connectionPoolClosed", () => {
    logger.warn("MongoDB closed")
    mongodbClient = null
  })

  await client.connect()
  // @ts-expect-error
  mongodbClientState = client.topology.s.state
  mongodbClient = client
  logger.info("Connected to MongoDB")

  return client
}

export const getMongodbClient = () => mongodbClient
export const getMongodbClientState = () => mongodbClientState

export const closeMongodbConnection = async () => {
  logger.warn("Closing MongoDB")
  // Let 100ms for possible callback cleanup to register tasks in mongodb queue
  await sleep(200)
  return mongodbClient?.close()
}

export const getDatabase = () => {
  return ensureInitialization().db()
}

export const getDbCollection = <K extends CollectionName>(name: K): Collection<IDocument<K>> => {
  return ensureInitialization().db().collection(name)
}

export const getCollectionList = () => {
  return ensureInitialization().db().listCollections().toArray()
}

export const getDbCollectionIndexes = async (name: CollectionName) => {
  return await ensureInitialization().db().collection(name).indexes()
}

/**
 * Création d'une collection si elle n'existe pas
 * @param {string} collectionName
 */
const createCollectionIfDoesNotExist = async (collectionName: string) => {
  const db = getDatabase()
  const collectionsInDb = await db.listCollections().toArray()
  const collectionExistsInDb = collectionsInDb.map(({ name }) => name).includes(collectionName)

  if (!collectionExistsInDb) {
    try {
      await db.createCollection(collectionName)
    } catch (err) {
      if ((err as MongoServerError).codeName !== "NamespaceExists") {
        throw err
      }
    }
  }
}

/**
 * Vérification de l'existence d'une collection à partir de la liste des collections
 * @param {*} collectionsInDb
 * @param {*} collectionName
 * @returns
 */
export const collectionExistInDb = (collectionsInDb: CollectionInfo[], collectionName: string) => collectionsInDb.map(({ name }: { name: string }) => name).includes(collectionName)

/**
 * Config de la validation
 * @param {*} modelDescriptors
 */
export const configureDbSchemaValidation = async (modelDescriptors: IModelDescriptor[]) => {
  const db = getDatabase()
  ensureInitialization()
  await Promise.all(
    modelDescriptors.map(async ({ collectionName, zod }) => {
      await createCollectionIfDoesNotExist(collectionName)

      const convertedSchema = zodToMongoSchema(zod)

      try {
        await db.command({
          collMod: collectionName,
          validationLevel: "strict",
          validationAction: "error",
          validator: {
            $jsonSchema: {
              title: `${collectionName} validation schema`,
              ...convertedSchema,
            },
          },
        })
      } catch (error) {
        captureException(error)
        logger.error(error)
      }
    })
  )
}

/**
 * Clear de toutes les collections
 * @returns
 */
export const clearAllCollections = async () => {
  const collections = await getDatabase().collections()
  return Promise.all(collections.map((c) => c.deleteMany({})))
}

/**
 * Clear d'une collection
 * @param {string} name
 * @returns
 */
export async function clearCollection(name: string) {
  ensureInitialization()
  await getDatabase().collection(name).deleteMany({})
}

export const createIndexes = async () => {
  for (const descriptor of modelDescriptors) {
    if (!descriptor.indexes) {
      return
    }
    const indexes = await getDbCollection(descriptor.collectionName)
      .listIndexes()
      .toArray()
      .catch((err) => {
        // NamespaceNotFound
        if (err.code === 26) {
          return []
        }
        throw err
      })
    const indexesToRemove = new Set(indexes.filter((i) => i.name !== "_id_"))

    logger.info(`Create indexes for collection ${descriptor.collectionName}`)
    await Promise.all(
      descriptor.indexes.map(async ([index, options]): Promise<void> => {
        try {
          const existingIndex =
            // Use Object.entries because order matters
            indexes.find((i) => isEqual(Object.entries(i.key), Object.entries(index)) || options.name === i.name) ?? null

          if (existingIndex) {
            indexesToRemove.delete(existingIndex)
          }

          await getDbCollection(descriptor.collectionName)
            .createIndex(index, options)
            .catch(async (err) => {
              // IndexOptionsConflict & IndexKeySpecsConflict
              if (err.code === 85 || err.code === 86) {
                await getDbCollection(descriptor.collectionName).dropIndex(existingIndex.name)
                await getDbCollection(descriptor.collectionName).createIndex(index, options)
              } else {
                throw err
              }
            })
        } catch (err) {
          captureException(err)
          logger.error(`Error creating indexes for ${descriptor.collectionName}: ${err}`)
        }
      })
    )

    if (indexesToRemove.size > 0) {
      logger.warn(`Dropping extra indexes for collection ${descriptor.collectionName}`, indexesToRemove)
      await Promise.all(Array.from(indexesToRemove).map((index) => getDbCollection(descriptor.collectionName).dropIndex(index.name)))
    }
  }
}

export const dropIndexes = async () => {
  const collections = (await getCollectionList()).map((collection) => collection.name)
  for (const descriptor of modelDescriptors) {
    logger.info(`Drop indexes for collection ${descriptor.collectionName}`)
    if (collections.includes(descriptor.collectionName)) {
      await getDbCollection(descriptor.collectionName).dropIndexes()
    }
  }
}
