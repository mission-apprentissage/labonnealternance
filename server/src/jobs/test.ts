import { ObjectId } from "bson"
import { IApplicant, IApplication, ZApplicant } from "shared"

import { getDbCollection } from "@/common/utils/mongodbUtils"

import { asyncForEach } from "../common/utils/asyncUtils"

type ApplicationAggregate = Pick<IApplication, "applicant_email" | "applicant_first_name" | "applicant_last_name" | "applicant_phone" | "_id">

export const migrationTest = async () => {
  // console.log(extensions.telephone.safeParse("0690760552"))
  // return
  const applications = (await getDbCollection("applications")
    .find({}, { projection: { applicant_first_name: 1, applicant_last_name: 1, applicant_email: 1, applicant_phone: 1, _id: 1 } })
    .toArray()) as ApplicationAggregate[]

  const stat = { error: 0, success: 0, total: applications.length }

  await asyncForEach(applications, async (application) => {
    const { applicant_first_name, applicant_email, applicant_last_name, applicant_phone, _id } = application
    const now = new Date()
    const applicant: IApplicant = {
      _id: new ObjectId(),
      firstname: applicant_first_name,
      lastname: applicant_last_name,
      phone: applicant_phone,
      email: applicant_email,
      last_connection: null,
      createdAt: now,
      updatedAt: now,
    }
    const validation = ZApplicant.safeParse(applicant)
    if (!validation.success) {
      stat.error++
      console.log({ applicant, err: validation.error.errors })
    } else {
      stat.success++
    }
  })
  console.log(stat)
}
