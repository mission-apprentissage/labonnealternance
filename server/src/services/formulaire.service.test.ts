import { useMongo } from "@tests/utils/mongo.test.utils"
import { saveEntrepriseUserTest } from "@tests/utils/user.test.utils"
import omit from "lodash/omit"
import { ObjectId } from "mongodb"
import { generateJobFixture } from "shared/fixtures/recruiter.fixture"
import { generateReferentielRome } from "shared/fixtures/rome.fixture"
import { IRecruiter, IReferentielRome, IUserWithAccount } from "shared/models"
import { beforeEach, describe, expect, it } from "vitest"

import { getDbCollection } from "@/common/utils/mongodbUtils"

import { createJob } from "./formulaire.service"

useMongo()

describe("createJob", () => {
  let user: IUserWithAccount
  let recruiter: IRecruiter
  let referentielRome: IReferentielRome

  beforeEach(async () => {
    const email = "entreprise@mail.fr"
    const response = await saveEntrepriseUserTest(
      {
        _id: new ObjectId("670ce1ded6ce30c3c90a0e1d"),
        email,
      },
      {},
      {},
      { jobs: [], email, _id: new ObjectId("670ce30b57a50d6875c141f9"), establishment_creation_date: new Date("2024-10-14T09:23:21.588Z") }
    )
    user = response.user
    recruiter = response.recruiter

    referentielRome = generateReferentielRome()
    await getDbCollection("referentielromes").insertOne(referentielRome)

    return async () => {
      await getDbCollection("userswithaccounts").deleteMany({})
      await getDbCollection("recruiters").deleteMany({})
      await getDbCollection("entreprises").deleteMany({})
      await getDbCollection("rolemanagements").deleteMany({})
      await getDbCollection("referentielromes").deleteMany({})
    }
  })

  const generateValidJobWritable = () => {
    return generateJobFixture({
      managed_by: user._id.toString(),
      rome_code: [referentielRome.rome.code_rome],
      competences_rome: {
        savoir_etre_professionnel: referentielRome.competences.savoir_etre_professionnel?.slice(0, 1),
        savoir_faire: referentielRome.competences.savoir_faire?.slice(0, 1),
        savoirs: referentielRome.competences.savoirs?.slice(0, 1),
      },
    })
  }

  it("should insert a job", async () => {
    const job = generateValidJobWritable()
    const result = await createJob({ user, establishment_id: recruiter.establishment_id, job })

    expect.soft(omit(result, "jobs")).toMatchSnapshot()
    expect.soft(result.jobs.length).toEqual(1)
    expect.soft(omit(result.jobs[0], "job_creation_date", "job_update_date", "_id", "job_expiration_date")).toMatchSnapshot()
  })
  it("should raise a bad request when savoir_etre_professionnel do not match referentiel rome", async () => {
    const job = generateValidJobWritable()
    job.competences_rome!.savoir_etre_professionnel = [
      {
        code_ogr: "test",
        libelle: "test",
        coeur_metier: "test",
      },
    ]
    expect.soft(() => createJob({ user, establishment_id: recruiter.establishment_id, job })).rejects.toThrow("compétences invalides")
  })
  it("should raise a bad request when savoir_faire do not match referentiel rome", async () => {
    const job = generateValidJobWritable()
    job.competences_rome!.savoir_faire = [
      {
        libelle: "test",
        items: [
          {
            code_ogr: "test",
            libelle: "test",
            coeur_metier: "test",
          },
        ],
      },
    ]
    expect.soft(() => createJob({ user, establishment_id: recruiter.establishment_id, job })).rejects.toThrow("compétences invalides")
  })
  it("should raise a bad request when savoirs do not match referentiel rome", async () => {
    const job = generateValidJobWritable()
    job.competences_rome!.savoirs = [
      {
        libelle: "test",
        items: [
          {
            code_ogr: "test",
            libelle: "test",
            coeur_metier: "test",
          },
        ],
      },
    ]
    expect.soft(() => createJob({ user, establishment_id: recruiter.establishment_id, job })).rejects.toThrow("compétences invalides")
  })
})
