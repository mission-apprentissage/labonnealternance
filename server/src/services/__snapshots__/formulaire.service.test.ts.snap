// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createJob > should insert a job 1`] = `
{
  "_id": "670ce30b57a50d6875c141f9",
  "address": "address",
  "address_detail": "address_detail",
  "cfa_delegated_siret": null,
  "createdAt": 2021-01-28T15:00:00.000Z,
  "distance": 10,
  "email": "entreprise@mail.fr",
  "establishment_creation_date": 2024-10-14T09:23:21.588Z,
  "establishment_enseigne": "establishment_enseigne",
  "establishment_id": "xxxx-xxxx-xxxx-xxxx",
  "establishment_raison_sociale": "establishment_raison_sociale",
  "establishment_siret": "wosdnnpetvgxonk",
  "establishment_size": "establishment_size",
  "first_name": "first_name",
  "geo_coordinates": "geo_coordinates",
  "geopoint": {
    "coordinates": [
      41,
      10,
    ],
    "type": "Point",
  },
  "idcc": "idcc",
  "is_delegated": false,
  "last_name": "last_name",
  "naf_code": "naf_code",
  "naf_label": "naf_label",
  "opco": "OPCO multiple",
  "origin": "origin",
  "phone": "phone",
  "status": "Actif",
  "updatedAt": 2021-02-03T17:00:00.000Z,
}
`;

exports[`createJob > should insert a job 2`] = `
{
  "competences_rome": {
    "savoir_etre_professionnel": [
      {
        "code_ogr": "300490",
        "coeur_metier": null,
        "libelle": "Faire preuve de rigueur et de précision",
      },
    ],
    "savoir_faire": [
      {
        "items": [
          {
            "code_ogr": "101622",
            "coeur_metier": "Principale",
            "libelle": "Procéder à l'enregistrement, au tri, à l'affranchissement du courrier",
          },
          {
            "code_ogr": "113818",
            "coeur_metier": null,
            "libelle": "Réaliser des travaux de reprographie",
          },
        ],
        "libelle": "Production, Fabrication",
      },
    ],
    "savoirs": [
      {
        "items": [
          {
            "code_ogr": "120992",
            "coeur_metier": null,
            "libelle": "Gestion administrative du courrier",
          },
        ],
        "libelle": "Domaines d'expertise",
      },
    ],
  },
  "is_disabled_elligible": null,
  "job_count": 1,
  "job_start_date": 2021-01-28T15:00:00.000Z,
  "job_status": "En attente",
  "job_type": [
    "Apprentissage",
  ],
  "managed_by": "670ce1ded6ce30c3c90a0e1d",
  "rome_code": [
    "M1602",
  ],
}
`;
